File: 500102181/Python/QuickSort.py
def Partition(lst, low, high):
    pivot = lst[low]
    i = low + 1
    j = high

    while True:
        while i <= j and lst[i] <= pivot:
            i += 1
        while i <= j and lst[j] >= pivot:
            j -= 1
        if i <= j:
            lst[i], lst[j] = lst[j], lst[i]
        else:
            break

    lst[low], lst[j] = lst[j], lst[low]
    return j

def QuickSort(lst, low, high):
    if low < high:
        partitionIndex = Partition(lst, low, high)
        QuickSort(lst, low, partitionIndex - 1)
        QuickSort(lst, partitionIndex + 1, high)

list = [2, 4, 3, 9, 1, 4, 8, 7, 5, 6]
n = len(list)
print("Unsorted list...\n")
print(list)
QuickSort(list, 0, n - 1)
print("Sorted list...\n")
print(list)

File: 500102181/quick_sort.c
#include <stdio.h>
#include <pbPlots.h>
#include <supportLib.h>

void print_array(int *arr, int n)
{
    for (int i = 0; i < n; i++)
        printf("%d\t", arr[i]);
}

int partition(int arr[], int low, int high)
{
    int pivot = arr[low];
    int i = low + 1;
    int j = high;
    int temp;
    do
    {
        while (arr[i] <= pivot)
            i++;
        while (arr[j] > pivot)
            j--;
        if (i < j)
        {
            temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    } while (i < j);
    temp = arr[low];
    arr[low] = arr[j];
    arr[j] = temp;
    return j;
}

void quick_sort(int arr[], int low, int high)
{
    int partitionIndex;
    if (low < high)
    {
        partitionIndex = partition(arr, low, high);
        quick_sort(arr, low, partitionIndex - 1);
        quick_sort(arr, partitionIndex + 1, high);
    }
}

int main()
{
    int n;
    printf("Enter size of array : ");
    scanf("%d", &n);
    int arr[n];
    printf("Enter %d elements in array : ", n);
    for (int k = 0; k < n; k++)
        scanf("%d", &arr[k]);
    printf("Unsorted array...\n");
    print_array(arr, n);
    printf("\n============================================\n");
    printf("Sorted array...\n");
    quick_sort(arr, 0, n - 1);
    print_array(arr, n);
    return 0;
}